### ISSUE TITLE
Use immutable API instead of getters and setters

### ISSUE THREAD

**Issue by jgaehring** [farmOS/farmOS.js#34]
I initially used a getter/setter pattern in the model to provide simpler ergonomics when updating entities without needing to think about metadata. For instance, to set the status on a log:

```js
log.status = 'done';
```

In the background, this also updates the `changed` metadata to the current timestamp for the purpose of merging remote changes. This was a change from the earlier `farmLog` implementation which used an `update()` function instead:

```js
const log2 = update(log1, { status: 'done' });
```

Although I thought this would help to simplify update operations, I'm learning in practice that this doesn't make a whole lot of difference, but it would be far more helpful to work with immutable data structures to prevent issues of shared mutable state.

This really wouldn't be too difficult, as it only requires moving the logic out of the getters/setters and returning the updated value, rather than mutating it. Eventually, this could be better for integrating validation into the update function. I believe this could enable better integration with the connect library, too, like performing merges with fetch/send operations, etc.

**Comment by jgaehring** [comment]
A few common immutable libraries for reference:

- React's [immutability-helper](https://github.com/kolodny/immutability-helper#update)
- Immer's [update patterns](https://immerjs.github.io/immer/update-patterns)

**Comment by jgaehring** [comment]
I feel like this would also eliminate some of the ugliness in creating entities from a remote that hasn't been synced to the local context before. In Field Kit, I have to manually pass the remote metadata in to the `create` method, which is kinda a pain:

```js
const now = new Date().toISOString();
const { meta } = remote;
const log = farm.log.create(remote, { remote: { lastSync: now, meta } });
```

**Comment by jgaehring** [comment]
> I have to manually pass the remote metadata in to the `create` method

It's actually slightly uglier, now that I realize `create` can't actually handle raw remotes from the server, even with `adapter.js` in the middle. Instead I have to use `deserialize`, with a lil help from Ramda:

```js
const now = new Date().toISOString();
const meta = m => ({ remote: { lastSync: now, meta: m } });
const serialized = evolve({ meta }, remote);
const log = farm.log.deserialize(serialized);
```

If `merge` could handle all of this, by accepting `undefined` as a first argument if there's no local entity and returning a brand new local version of the remote, that would be soooo much cleaner. But right now `merge` acts on the local entity _in place_, and has no return value. I feel like the whole mental model would become more straightforward with immutability.

**Comment by jgaehring** [comment]
> Instead I have to use `deserialize`

Another thing: I could eliminate the `serialize` and `deserialize` methods.

**Comment by jgaehring** [comment]
Resolved by 9019f6f.
